File: Java/TreePractice/src/bst/BinarySearchTreeTest.java
package bst;

import node.TreeNode;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.io.ByteArrayOutputStream;
import java.io.PrintStream;

import static org.junit.jupiter.api.Assertions.*;

class BinarySearchTreeTest {

    private final PrintStream standardOut = System.out;
    private final ByteArrayOutputStream outputStreamCaptor = new ByteArrayOutputStream();

    TreeNode jessica = new TreeNode("jessica");

    BinarySearchTree binarySearchTree;

    @BeforeEach
    void setUp() {
        System.setOut(new PrintStream(outputStreamCaptor));
        this.binarySearchTree = new BinarySearchTree(this.jessica);
        binarySearchTree.insert("alice");
        binarySearchTree.insert("ramona");
        binarySearchTree.insert("laurie");
        binarySearchTree.insert("xiaohan");
        binarySearchTree.insert("bert");
        binarySearchTree.insert("joshua");
    }

    @AfterEach
    void tearDown() {
        System.setOut(standardOut);
    }

    @Test
    void getRoot() {
        assertEquals(this.binarySearchTree.getRoot(), this.jessica);
    }

    @Test
    void insert() {
        this.binarySearchTree.insert("donald");
        assertNotNull(this.binarySearchTree.lookup("donald"));
    }

    @Test
    void lookup() {
        TreeNode result = (TreeNode) this.binarySearchTree.lookup("jessica");
        assertEquals(result.getData(), "jessica");
        TreeNode result2 = (TreeNode) this.binarySearchTree.lookup("bert");
        assertEquals(result2.getData(), "bert");
    }

    @Test
    void printAll() {
        this.binarySearchTree.printAll();
        String output = outputStreamCaptor.toString().trim();
        assertTrue(output.contains("alice"));
        assertTrue(output.contains("ramona"));
        assertTrue(output.contains("laurie"));
        assertTrue(output.contains("xiaohan"));
        assertTrue(output.contains("bert"));
        assertTrue(output.contains("joshua"));
    }

    @Test
    void remove() {
        this.binarySearchTree.remove("bert");
        TreeNode result = (TreeNode) this.binarySearchTree.lookup("bert");
        assertNull(result);
    }
}

File: Kotlin/TreePractice/test/bst/BinarySearchTreeTest.kt
package bst

import org.junit.jupiter.api.Test
import org.junit.jupiter.api.BeforeEach
import java.io.ByteArrayOutputStream
import java.io.PrintStream

import org.junit.jupiter.api.Assertions.*

class BinarySearchTreeTest {

    private val outputStreamCaptor = ByteArrayOutputStream()

    private val binarySearchTree = BinarySearchTree()

    @BeforeEach
    fun setUp() {
        System.setOut(PrintStream(outputStreamCaptor))
        binarySearchTree.insert("F")
        binarySearchTree.insert("S")
        binarySearchTree.insert("C")
        binarySearchTree.insert("H")
        binarySearchTree.insert("V")
        binarySearchTree.insert("A")
        binarySearchTree.insert("M")
    }

    @Test
    fun insert() {
        binarySearchTree.insert("M")
        binarySearchTree.lookup("M")?.let { assertEquals(it.data, "M") }
    }

    @Test
    fun lookup() {
        binarySearchTree.lookup("C")?.let { assertEquals(it.data, "C") }
        binarySearchTree.lookup("X")?.let { assertEquals(it.data, null) }
    }

    @Test
    fun getRoot() {
        binarySearchTree.getRoot()?.let { assertEquals(it, "F")}
    }

    @Test
    fun printAll() {
        this.binarySearchTree.printAll()
        val output = outputStreamCaptor.toString().trim { it <= ' ' }
        assertTrue(output.contains("F"))
        assertTrue(output.contains("S"))
        assertTrue(output.contains("C"))
        assertTrue(output.contains("H"))
        assertTrue(output.contains("V"))
        assertTrue(output.contains("A"))
        assertTrue(output.contains("M"))
    }

    @Test
    fun remove() {
        binarySearchTree.remove("F")
        binarySearchTree.lookup("F")?.let { assertEquals(it.data, null) }
        binarySearchTree.getRoot()?.let { assertEquals(it, "H")}
        binarySearchTree.getRoot()?.let { assertNotEquals(it, "F")}
        binarySearchTree.remove("S")
        binarySearchTree.lookup("S")?.let { assertEquals(it.data, null) }
    }
}